### PR TITLE
Fix/refactor pfba

### PR DESCRIPTION
**Pull request by hredestig**
adds the add_pfba function and adjusts the optimize_minimal_flux to use
either optlang implementation if requested

### CODE REVIEW

File: cobra/flux_analysis/parsimonious.py [hunk 1]
```diff
@@ -1,22 +1,151 @@
...
+            solver=solver,
+            desired_objective_value=desired_objective_value,
+            **optimize_kwargs)
+
+
+def add_pfba(model, objective=None, fraction_of_optimum=1.0):
+    if objective is not None:
+        model.objective = objective
+    model.fix_objective_as_constraint(fraction=fraction_of_optimum)
+    pfba_objective = model.solver.interface.Objective(add(
```
Comment (cdiener) [inline cobra/flux_analysis/parsimonious.py]: This should only include variables belonging to reactions. There might be other variables defined in the solver that should not be minimized.

File: cobra/core/Model.py [hunk 2]
```diff
@@ -643,3 +645,35 @@ def __exit__(self, type, value, traceback):
         """Pop the top context manager and trigger the undo functions"""
         context = self._contexts.pop()
         context.reset()
+
+    def fix_objective_as_constraint(self, fraction=1):
```
Comment (cdiener) [inline cobra/core/Model.py]: Maybe put that in until.solver. Seems useful.

File: cobra/flux_analysis/parsimonious.py (hunk 1 above)
Comment (hredestig) [inline cobra/flux_analysis/parsimonious.py]: Good point, fixed

**Comment by hredestig** [comment]
```
Name (time in ms)                            Mean          
-----------------------------------------------------------
test_pfba_benchmark[cplex]               346.9930 (1.0)    
test_pfba_benchmark[optlang-cplex]       412.9751 (1.19)   
test_pfba_benchmark[optlang-glpk]        398.0229 (1.15)   
test_pfba_benchmark[cglpk]               477.0672 (1.37)   
-----------------------------------------------------------
```
suprised old cplex is better wonder if it's really correct...

File: cobra/flux_analysis/parsimonious.py [hunk 3]
```diff
@@ -1,22 +1,174 @@
...
+        Fraction of optimum which must be maintained. The original objective
+        reaction is constrained to be greater than maximal_value *
+        fraction_of_optimum.
+    """
+    if objective is not None:
+        model.objective = objective
+    sutil.fix_objective_as_constraint(model, fraction=fraction_of_optimum)
+    reaction_variables = ((rxn.forward_variable, rxn.reverse_variable)
+                          for rxn in model.reactions)
+    variables = chain(*reaction_variables)
```
Comment (cdiener) [inline cobra/flux_analysis/parsimonious.py]: Nice, didn't know about chain :O :+1:

**Pull request merged**